### ISSUE TITLE
Some confusion about the grad of torch.sign() 

### ISSUE THREAD

**Issue by jxgu1016** [pytorch/pytorch#7940]
Here's my confusion on [pytorch forums discussion](https://discuss.pytorch.org/t/some-confusion-about-the-grad-of-torch-sign/18762). 

Still confused and need more attention.

**Comment by SsnL** [comment]
You should call `.retain_grad`

**Comment by soumith** [comment]
@SsnL once we call `requires_grad_()`, grad should become non-None, it has to effectively become a leaf-node. Isn't it the purpose why we introduced `requires_grad_()`, so that a node that wasn't requiring grad, can suddenly be converted to requiring it.

**Comment by soumith** [comment]
cc: @gchanan

**Comment by SsnL** [comment]
@soumith Not really. The purpose of `requires_grad_()` is mainly to replace the pattern `Variable(x.data, requires_grad=True)`. It should be used as `x.data.requires_grad_()` or `x.detach().requires_grad_()`. `requires_grad_` it self only sets the flag, and doesn't detach. (That's why you can do things like `requries_grad_(False)`.)

**Comment by soumith** [comment]
oh I see, so `x.detach().requires_grad_()` is the right usage here then.
